src/Surface.mo

```diff
@@ -0,0 +1,311 @@
+/// Motoko surface syntax, as a recursive type (a tree).
+///
+module {
+
+  type Pos = {
+    line : Nat;
+    col : Nat;
+  };
+
+  type Region = {
+    begin : Pos;
+    end : Pos;
+  };
+
+  /// In a multi-user setting, an expression may "accumulate" authors.
+  /// (Consider Motoko Playground, but in a multi-user mode.)
+  type Authors = [Principal];
+
+  type Annotated<X> = {
+    file : Text;
+    region : Region;
+    authors : ?Authors;
+    it : X;
+  };
+  
+  /// Expressions whose sub-expressions are generically typed.
+  type Exp_<E> = {
+    #prim : Text;
+    #var_ : Id;
+    #lit : Lit;
+    #actorUrl : E;
+    #un : (UnOp, E);
+    // to do -- continue.
+  };
+
+  /// Plain expressions, without annotations ("plain ASTs").
+  type Exp = Exp_<Exp>; 
+
+  /// Expressions where source-location annotations are present throughout.
+  type ExpA = Annotated<Exp_<ExpA>>;
```

Comment:
@rvanasa Using `mo_defs/surface.ml` as the reference presents a question: Should we really include all of these annotations in this more "mobile" representation of programs?

Some information in the annotations is relevant in the mobile setting, like how the AST node originated from a source filesystem, and who authored those files.  Other information is just for the compiler to use as scratch space. 


For the former information, my feeling is that sometimes we want this in mobile or dynamic programs, but other times we do not.

To accommodate both situations without duplicating the definition of each case, I've tried the pattern above.

WDYT?

Reply:
This seems like a very elegant solution which could also translate well to the "sugared" OCaml AST.